src/csharp/Grpc.Tools.Tests/MsBuildIntegrationTest.cs

```diff
@@ -0,0 +1,382 @@
+            string testId = Guid.NewGuid().ToString();
+
+            TryRunMsBuild("TestSingleProto",
+                "file.proto:File.cs;FileGrpc.cs",
+                testId
+                );
+        }
+
+        [Test]
+        public void TestMultipleProtos()
+        {
+            SkipIfMonoOrNet45();
+            string testId = Guid.NewGuid().ToString();
+
+            TryRunMsBuild("TestMultipleProtos",
+                "file.proto:File.cs;FileGrpc.cs" +
+                "|protos/another.proto:Another.cs;AnotherGrpc.cs" +
+                "|second.proto:Second.cs;SecondGrpc.cs",
+                testId
+                );
+        }
+
+        [Test]
+        public void TestAtInPath()
+        {
+            SkipIfMonoOrNet45();
+            string testId = Guid.NewGuid().ToString();
+
+            TryRunMsBuild("TestAtInPath",
+                "@protos/file.proto:File.cs;FileGrpc.cs",
+                testId
+                );
+        }
+
+        [Test]
+        public void TestProtoOutsideProject()
+        {
+            SkipIfMonoOrNet45();
+            string testId = Guid.NewGuid().ToString();
+
+            TryRunMsBuild("TestProtoOutsideProject/project",
+                "../api/greet.proto:Greet.cs;GreetGrpc.cs",
+                testId
+                );
+        }
+
+        /// <summary>
+        /// Run "dotnet build" on the test's project file.
+        /// </summary>
+        /// <param name="testName">Name of test and name of directory containing the test</param>
+        /// <param name="filesToGenerate">Tell the fake protoc script which files to generate</param>
+        /// <param name="testId">A unique ID for the test run - used to create results file</param>
+        private void TryRunMsBuild(string testName, string filesToGenerate, string testId)
+        {
+            var assemblyDir = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
+            var parentDir = System.IO.Directory.GetParent(assemblyDir).FullName;
+
+            // Path for fake proto
+            var fakeProtocScript = Platform.IsWindows ? "fakeprotoc.bat" : "fakeprotoc.py";
+            var fakeProtoc = Path.GetFullPath($"{parentDir}/../../scripts/{fakeProtocScript}");
+
+            // Paths for Grpc.Tools files
+            var grpcToolsDir = Path.GetFullPath($"{parentDir}/../../../Grpc.Tools");
+            var grpcToolsBuildDir = Path.GetFullPath($"{grpcToolsDir}/build");
+            // Task assembly is needed to run the extension tasks
+            var tasksAssembly = Path.GetFullPath($"{grpcToolsDir}/bin/Debug/netstandard1.3/{TASKS_ASSEMBLY_DLL}");
+
+            // Paths for test data
+            var testdataDir = Path.GetFullPath(parentDir+ "/../../Integration.Tests/");
+            var testDir = Path.GetFullPath(testdataDir + testName);
+
+            Console.WriteLine($"testDir = {testDir}");
+
+            // Clean up test output dirs: bin obj log
+            if (Directory.Exists(testDir+"/bin"))
```

Comment:
alternatively, you could get rid of the DeleteDirectoryWithRetry and issues with parallel runs by simply generating a 
"invocation_UUID" directory for each run and putting bin/ obj/ and log/ directories under it. That way each run of the test will be completely independent and you won't have the inconsistently of having a UUID.json file with testId in the name, but msbuild.log and fakeprotoc-dbg.txt without it.